lib/segment/src/index/query_optimization/nested_filter.rs

```diff
@@ -0,0 +1,373 @@
+pub fn find_indices_matching_all_conditions(
+    point_id: PointOffsetType,
+    nested_checkers: &Vec<NestedMatchingIndicesFn>,
+) -> Vec<ElemIndex> {
+    let condition_len = nested_checkers.len();
+    let mut matches: HashMap<ElemIndex, usize> = HashMap::new();
+    for f in nested_checkers {
+        let inner_matches = f(point_id);
+        for inner in inner_matches {
+            let count = matches.entry(inner).or_insert(0);
+            *count += 1;
+        }
+    }
+    // gather all indices that have matched all musts conditions
+    matches
+        .iter()
+        .filter(|(_, &count)| count == condition_len)
+        .map(|(index, _)| *index)
+        .collect()
+}
+
+pub fn nested_conditions_converter<'a>(
+    conditions: &'a [Condition],
+    field_indexes: &'a IndexesMap,
+    payload_provider: PayloadProvider,
+    nested_path: JsonPathPayload,
+) -> Vec<NestedMatchingIndicesFn<'a>> {
+    let mut nested_checker_fns = vec![];
+    conditions.iter().for_each(|condition| {
+        let condition_checker = nested_condition_converter(
+            condition,
+            field_indexes,
+            payload_provider.clone(),
+            nested_path.clone(),
+        );
+        nested_checker_fns.push(condition_checker);
+    });
+    nested_checker_fns
+}
+
+pub fn nested_condition_converter<'a>(
+    condition: &'a Condition,
+    field_indexes: &'a IndexesMap,
+    payload_provider: PayloadProvider,
+    nested_path: JsonPathPayload,
+) -> NestedMatchingIndicesFn<'a> {
+    match condition {
+        Condition::Field(field_condition) => {
+            let full_path = nested_path.extend(&field_condition.key);
+            field_indexes
+                .get(&full_path.path)
+                .and_then(|indexes| {
+                    indexes
+                        .iter()
+                        .filter_map(|index| nested_field_condition_index(index, field_condition))
+                        .next()
+                })
+                .unwrap_or_else(|| {
+                    Box::new(move |point_id| {
+                        payload_provider.with_payload(point_id, |payload| {
+                            nested_check_field_condition(field_condition, &payload, &nested_path)
+                        })
+                    })
+                })
+        }
+        Condition::IsEmpty(is_empty) => Box::new(move |point_id| {
+            payload_provider.with_payload(point_id, |payload| {
+                check_nested_is_empty_condition(&nested_path, is_empty, &payload)
+            })
+        }),
+        Condition::IsNull(is_null) => Box::new(move |point_id| {
+            payload_provider.with_payload(point_id, |payload| {
+                check_nested_is_null_condition(&nested_path, is_null, &payload)
+            })
+        }),
+        Condition::HasId(_) => {
+            // No support for has_id in nested queries
```

Comment:
Good question!

There are basically two reasons:

First in terms of use cases, this matches points by id so it is not related to the payload.
For a user, it can be added as an individual condition within the outer block to match on a specific point id.

Secondly I could not get it to work for the nested object filter implementation.

It requires access to the `id_tracker` which is borrowed to build the closure returned.
However, in case of nested the closure is called recursively and the `id_tracker` is now part of it.
It caused lifetime issues because the closure is called after the `id_tracker` is dropped.

I did not want to change more things to make it work as I was not sure how to tackle this.